### PR TITLE
Ing fork ph for leaves

### PR DESCRIPTION
**Pull request by alexeykoren**
This one consists mostly of custom embedded franklin-crypto, now we use it as sources folder rather than remote dependency. Doesn't look too pretty but less maintenance overhead, we have too many forks already.

The only chnage that is made in franklin-cypto is the ability to use variable number of PH generators.

### CODE REVIEW

**Review by mvdbos — APPROVED** [review]
Not really liking the full inclusion of franklin-crypto, but seems we will not be able to avoid it.

File: zinc-vm/src/lib.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub trait Engine: JubjubEngine + Debug {
 }
 
 lazy_static! {
-    static ref JUBJUB_BN256_PARAMS: AltJubjubBn256 = AltJubjubBn256::new_with_custom_generators(10);
+    static ref JUBJUB_BN256_PARAMS: AltJubjubBn256 = AltJubjubBn256::new_with_custom_generators(4 * 8 * 20); // ~ 20 Kb
```
[the remark this answers is not in the document]
Comment (alexeykoren) [reply zinc-vm/src/lib.rs:24]: number of elliptic curve points

**Comment by alexeykoren** [comment]
> Not really liking the full inclusion of franklin-crypto, but seems we will not be able to avoid it. Perhaps we could do it as a patch?

IIRC we've discussed it back then and decided that this is lesser evil. Alternative is to make franklin-crypto repo for our fork and use it as dependency.

**Comment by mvdbos** [comment]
> IIRC we've discussed it back then and decided that this is lesser evil. Alternative is to make franklin-crypto repo for our fork and use it as dependency.

I recall that as well. That is why I approved the PR anyway. :-) Just thinking out loud on perhaps using a patch instead of full copy, but not something for this moment. No confusion intended.

**Pull request merged**